### DISCUSSION TITLE
Creating plugins

### DISCUSSION THREAD

**Discussion by wobsoriano** [nuxt/nuxt#18344] [Questions] [answered]
Hi, the way we create plugins in Nuxt 2 is like this:

```ts
export default defineNuxtPlugin((context, inject) => {
  context.$hello = msg => `Hello ${msg}!`
  inject('hello', msg => `Hello ${msg}!`)
}
```

What's the equivalent of this in Nuxt 3? I tried this but not sure if it's the right way

```ts
export default defineNuxtPlugin((nuxt) => {
  // not sure how to access this in script but I can access in template
  nuxt.provide('hello', msg => `Hello ${msg}!`)
})
```

or via `provide/inject`?

```ts
export default defineNuxtPlugin(({ app }) => {
   app.provide('hello', msg => `Hello ${msg}!`)
})
```

```html
<script setup lang="ts">
const hello = inject('hello')

console.log(hello('robert'))
</script>
```

**Comment by r9mp** [comment]
Hi,
You can still do the same. Docs have been added on this subject : https://v3.nuxtjs.org/docs/directory-structure/plugins
So, just create a "plugins/" directory.
And then create your plugin inside in the exact same way than for the V2.
Everything will be auto detected. Just take car to add ".js" and ".vue" extensions to your imports ; Vite doesn't auto look for extensions.

**Reply by wobsoriano** [reply]
Hi yeah that's what I did on my example above, but the way we do it before is by attaching it to the `context` or via the 2nd param `inject`

```ts
export default defineNuxtPlugin((context, inject) => {
  context.$hello = msg => `Hello ${msg}!` // we can use context._legacyContext for this
  inject('hello', msg => `Hello ${msg}!`) // we don't have this in nuxt 3
}
```

just wondering if my implementation is correct (via `provide/inject`)

**Comment by pi0** [comment, marked as answer]
Hi. Sorry for lack of documentation. (wip: #884). Actually first argument passed to `defineNuxtPlugin`, unlike nuxt2 is `nuxtApp` and has provide utility. Example: (https://stackblitz.com/edit/github-rqdccs)

```js
import { defineNuxtPlugin } from '#app'

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.provide('hello', (msg) => `Hello ${msg}!`);
})
```

Alternative: (using composable to access nuxtApp)

```js
import { defineNuxtPlugin, useNuxtApp } from '#app';

export default defineNuxtPlugin((nuxtApp) => {
  provideHello();
});

function provideHello() {
  const nuxtApp = useNuxtApp();
  nuxtApp.provide('hello', (msg) => `Hello ${msg}!`);
}
```

**Reply by wobsoriano** [reply]
Thank you!
